### PR TITLE
Provide implementation of BQIO.Read that does not use Source API

### PR DESCRIPTION
**Pull request by sammcveety**
@tgroh (and other interested parties) does this match your understanding of what needs to be built, with respect to structure?  Is the code reuse of *Source reasonable?

@dhalperi FYI

### CODE REVIEW

**Comment by dhalperi** [comment]
@jkff also wants to see this, as it's "SDF-like" (assuming it mirrors whiteboard discussion -- haven't looked)

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 1]
```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
...
+                      List<String> files = source.extractFiles(p.getOptions());
+                      for (String file : files) {
+                        c.output(file);
+                      }
+                    }
+                  }));
+          final PCollectionView<TableSchema> schemaView = tuple.get(tableSchemaTag)
+              .apply(View.<TableSchema>asSingleton());
+          rows = tuple.get(filesTag)
+              //.apply(Reshuffle.<String, String>of())
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The reshuffle needs to stay, I think.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 2]
```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
...
+                    public void processElement(ProcessContext c)
+                        throws Exception {
+                      TableSchema schema = c.sideInput(schemaView);
+                      BigQuerySourceBase source = applySourceTransform(p, null, null);
+                      List<BoundedSource<TableRow>> sources =
+                          source.createSources(ImmutableList.of(c.element()), schema);
+                      for (BoundedSource<TableRow> avroSource : sources) {
+                        BoundedSource.BoundedReader<TableRow> reader =
+                            avroSource.createReader(p.getOptions());
+                        if (reader.start()) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The common idiom is: for (bool more = reader.start(); more; more = reader.advance())

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 3]
```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
...
+              //.apply(Reshuffle.<String, String>of())
+              .apply("ReadFiles", ParDo.of(
+                  new DoFn<String, TableRow>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext c)
+                        throws Exception {
+                      TableSchema schema = c.sideInput(schemaView);
+                      BigQuerySourceBase source = applySourceTransform(p, null, null);
+                      List<BoundedSource<TableRow>> sources =
+                          source.createSources(ImmutableList.of(c.element()), schema);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This should be a ParDo as well, and with a reshuffle, if there can be several of these.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 4]
```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
...
+                  public String apply(String input) {
+                    return randomUUIDString();
+                  }
+                }));
+          jobIdTokenView = jobIdTokenCollection
+              .apply(View.<String>asSingleton());
         }
+
+        PCollection<TableRow> rows;
+        if (true) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: What will the condition actually be?.. The non-Source version, without SDF, will not have dynamic rebalancing, so it better be a condition that justifies that :)

Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Initially, I was just thinking an experiment flag.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 3 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Several sources, right?  Should only be one, since we're only supplying one file (though that is not at all clear from the code).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 5]
```diff
@@ -715,49 +667,136 @@ public void validate(PBegin input) {
         }
       }
 
+      private BigQuerySourceBase applySourceTransform(
+          Pipeline p, String jobUuid, String extractDestinationDir) {
+        BigQuerySourceBase source;
+        BigQueryOptions bqOptions = p.getOptions().as(BigQueryOptions.class);
+        final BigQueryServices bqServices = getBigQueryServices();
+        final String executingProject = bqOptions.getProject();
```
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I'm generally leery of using the options that are available at application time, mainly because I'm trying to get rid of the ability to access them. The graph in the runner API doesn't have options associated with it, so if it's not being explicitly passed it probably shouldn't be used in the `expand()`. Can the `QuerySource` construct its own temp ref in `splitIntoBundles` or `createReader`?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I believe what is being suggested is that we can just output all of the created sources, and have a `DoFn` that reads from the `BoundedSource` we produce - and that way the runner can determine how to distribute that work.

Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Just to be sure I'm on the same page before I move forward with a rewrite, are we suggesting (or want to discuss further):

- DoFn 1: JobId (String) -> Sources (BoundedSource<TableRow>)
- DoFn 2: Source (BoundSource<TableRow>) -> Records (TableRow)

@tgroh @jkff can you explicitly ack?

**Comment by dhalperi** [comment]
R: @dhalperi

**Comment by dhalperi** [comment]
Will TAL today

**Comment by LarsOL** [comment]
Does this address the same as #2657 (Generate BigQuery load job at runtime)? 
Since that has been merged.

**Comment by sammcveety** [comment]
That applies to Write; this applies to Read.

**Comment by sammcveety** [comment]
@jkff I think this is almost ready for review.  One question (maybe for @kennknowles ), how would you suggest pluming useNewSource now that Pipeline doesn't have getOptions()?  I'd imagine one could do some template-specific handling of BQIO.Read (call setUseNewSource, or something), but wanted to get your thoughts.

**Comment by kennknowles** [comment]
In truth, I'm not opposed to a mechanism _like_ `PipelineOptions` to provide global settings during pipeline construction, but that won't actually solve the problem. The underlying issue is that templates already can't control expansions of transforms (I may be out of date) because the full pipeline is constructed prior to dynamic options being provided.

We got into trouble using `PipelineOptions` to control expansion because of the abstraction perforation inherent in `PipelineOptions.as(...)`: transforms could observe runner options and vice versa and each could alter their behavior in non-portable ways. I'm belaboring this in part to remind myself of why we cannot reintroduce it as-is. @lukecwik - can `PipelineOptions` support package private options? Other ideas? But I digress - this still won't let a template control expansion.

**Comment by lukecwik** [comment]
@kennknowles No, from [javadoc](https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/Proxy.html#getProxyClass(java.lang.ClassLoader,%20java.lang.Class...)): "All non-public interfaces must be in the same package; otherwise, it would not be possible for the proxy class to implement all of the interfaces, regardless of what package it is defined in."

We might be able to do something with ByteBuddy or ASM to remove this limitation but a significant rewrite of `PipelineOptionsFactory`/`ProxyInvocationHandler` would be required.

**Comment by sammcveety** [comment]
@kennknowles I may be misunderstanding, but I think the specific case I'm
referring to (branching on what source implementation to use) would have
solved the problem because the selection of "I am creating a template" (and
a corresponding flip on a BQ-specific toggle) is available at construction
time.  The specific parameters to the source (e.g. table) would be dynamic
and inaccessible, but that's orthogonal.

I'm rusty on the I/O substitution stuff we have around for native sources,
but maybe there's a hook there I could use?

**Comment by kennknowles** [comment]
@sammcveety OK, I thought you wanted the template instantiator to control this. In that case, the problem is the one I mused upon. I don't have a ready solution that let's `BigQueryIO` own the decision. A runner has access to options and could do a substitution, if that is what you are alluding to, but it would be owned by the runner.

**Comment by sammcveety** [comment]
@jkff ready for review

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 6]
```diff
@@ -397,11 +414,43 @@ public Read usingStandardSql() {
...
+    public Read withNewSource() {
+      return toBuilder().setUseNewSource(true).build();
+    }
+
     @VisibleForTesting
     Read withTestServices(BigQueryServices testServices) {
       return toBuilder().setBigQueryServices(testServices).build();
     }
 
+    private BigQuerySourceBase applySourceTransform(String jobUuid) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This does not apply a transform - perhaps just call this createSource?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 7]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
     @Override
     public PCollection<TableRow> expand(PBegin input) {
-      final String stepUuid = BigQueryHelpers.randomUUIDString();
-      BoundedSource<TableRow> source;
+      Pipeline p = input.getPipeline();
+      final String staticJobUuid = BigQueryHelpers.randomUUIDString();
+
+      final PCollectionView<String> jobIdTokenView;
+      PCollection<String> jobIdTokenCollection = null;
+      if (!getUseNewSource()) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Can you merge this if and the next one at line 561, and then move staticJobUuid into the "if" branch?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 8]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
+            .setCoder(getDefaultOutputCoder());
       } else {
-        source =
-            BigQueryTableSource.create(
-                stepUuid,
-                getTableProvider(),
-                getBigQueryServices());
+        // Apply the DoFn version.
+        final TupleTag<String> filesTag =
+            new TupleTag<String>(){};
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: `{}` is probably not necessary here, and probably harmful - it captures an instance of the enclosing PTransform and blows up the serialized graph. It only makes the TupleTag remember its type (String), but that's not necessary - if you get a coder inference error, you can just `setCoder` somewhere below until it disappears.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 9]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
+            .apply(Reshuffle.<String, String>of())
+            .apply(Values.<String>create())
+            .apply("ReadFiles", ParDo.of(
+                new DoFn<String, TableRow>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    TableSchema schema = BigQueryHelpers.fromJsonString(
+                        c.sideInput(schemaView), TableSchema.class);
+                    String jobUuid = (String) c.sideInput(jobIdTokenView);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The cast should be unnecessary

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 10]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
+                    BigQuerySourceBase source = applySourceTransform(jobUuid);
+                    List<BoundedSource<TableRow>> sources =
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This should in practice create exactly 1 element, right? I think it'd make the code more readable if you enforced that.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 11]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
+                    source.createSources(ImmutableList.of(
+                        FileSystems.matchNewResource(
+                            c.element(), false /* is directory */)), schema);
+                    for (BoundedSource<TableRow> avroSource : sources) {
+                      BoundedSource.BoundedReader<TableRow> reader =
+                          avroSource.createReader(c.getPipelineOptions());
+                      if (reader.start()) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: The more common idiom for this is: `for (boolean more = reader.start(); more; more = reader.advance())`

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 12]
```diff
@@ -483,41 +532,115 @@ public void validate(PipelineOptions options) {
...
+                      }
+                    }
+                  }
+                }).withSideInputs(schemaView, jobIdTokenView));
       }
       PassThroughThenCleanup.CleanupOperation cleanupOperation =
           new PassThroughThenCleanup.CleanupOperation() {
             @Override
-            void cleanup(PipelineOptions options) throws Exception {
+            void cleanup(DoFn.ProcessContext c) throws Exception {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This probably shouldn't be a raw type. Then you can remove the raw cast below.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java [hunk 13]
```diff
@@ -79,6 +81,6 @@ public void processElement(ProcessContext c) {
   }
 
   abstract static class CleanupOperation implements Serializable {
-    abstract void cleanup(PipelineOptions options) throws Exception;
+    abstract void cleanup(DoFn.ProcessContext options) throws Exception;
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: Should probably not take a raw type, nor should it take a DoFn.ProcessContext in general, because it shouldn't be allowed to do anything with the context except get the pipeline options and have access to a side input. Consider instead mimicking (or extracting) the pattern from https://github.com/apache/beam/blob/master/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.java#L76 - which conveniently enough is also in BigQuery.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 12 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Not sure I'm getting this right -- if I want to remove the sideInput cast, I would type that call, but that seems orthogonal to the DoFn<> typing.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java (hunk 13 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: Something like this?

class JobIdContainer {

public JobIdContainer(DoFn<?, ?>.ProcessContext pc, PCView<> view) {
this.view = ...
this.pc = ...
}

public getJobId() {
return …
}

}

Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: Yes, this seems reasonable, assuming job id is the only thing to which you want CleanupOperation to have access (rather than arbitrary side inputs).

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 14]
```diff
@@ -397,11 +414,43 @@ public Read usingStandardSql() {
       return toBuilder().setUseLegacySql(false).build();
     }
 
+    /**
+     * Use new SplittableDoFn-compatible source implementation.
```
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 6 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 7 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 8 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Could you point me to where I want setCoder?  MultiOutput doesn't have coder methods, and the docs appear to advocate for "{}": https://beam.apache.org/documentation/programming-guide/#transforms-sideio

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 9 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Still grumps at me after moving:

ThroughThenCleanup.java:[103,39] incompatible types: java.lang.Object cannot be converted to java.lang.String

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 10 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java (hunk 11 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java (hunk 13 above)
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java [hunk 15]
```diff
@@ -41,9 +40,12 @@
 class PassThroughThenCleanup<T> extends PTransform<PCollection<T>, PCollection<T>> {
 
   private CleanupOperation cleanupOperation;
+  private PCollectionView<String> sideInput;
 
-  PassThroughThenCleanup(CleanupOperation cleanupOperation) {
+  PassThroughThenCleanup(CleanupOperation cleanupOperation,
+      PCollectionView<String> sideInput) {
```
Comment (sammcveety) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: Done.

**Comment by sammcveety** [comment]
retest this please

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 16]
```diff
@@ -397,11 +414,44 @@ public Read usingStandardSql() {
...
 
+    /**
+     * Use new template-compatible source implementation.
+     *
+     * <p>Use new template-compatible source implementation. This implementation is
+     * compatible with repeated template invocations. It does not support dynamic work
+     * rebalancing.
+     */
+    @Experimental(Experimental.Kind.SOURCE_SINK)
+    public Read withNewSource() {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: I still think that "withNewSource" is not a good name for this method: if any user that wants to use BigQueryIO.read() with templates must call this method, then the method should be called withTemplateCompatibility() or something like that.

Also, would be good if the comment explained why we can't automatically call this if the job is being run from a template and why the user has to do it instead.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 17]
```diff
@@ -397,11 +414,44 @@ public Read usingStandardSql() {
...
+
     @VisibleForTesting
     Read withTestServices(BigQueryServices testServices) {
       return toBuilder().setBigQueryServices(testServices).build();
     }
 
+    private BigQuerySourceBase createSource(String jobUuid) {
+      BigQuerySourceBase source;
+      if (getQuery() != null
+          && (!getQuery().isAccessible() || !Strings.isNullOrEmpty(getQuery().get()))) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: This condition would read a lot better if inverted

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 18]
```diff
@@ -483,41 +533,107 @@ public void validate(PipelineOptions options) {
...
+                    String schema = BigQueryHelpers.toJsonString(
+                        source.getSchema(c.getPipelineOptions()));
+                    c.output(tableSchemaTag, schema);
+                    List<ResourceId> files = source.extractFiles(c.getPipelineOptions());
+                    for (ResourceId file : files) {
+                      c.output(file.toString());
+                    }
+                  }
+                })
+            .withOutputTags(filesTag, TupleTagList.of(tableSchemaTag)));
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: By coders, I meant:
- Remove `{}` at lines 565, 567
- Here, say: `tuple.get(filesTag).setCoder(StringUtf8Coder.of()); tuple.get(tableSchemaTag).setCoder(StringUtf8Coder.of());`

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 19]
```diff
@@ -483,41 +533,107 @@ public void validate(PipelineOptions options) {
...
+                new DoFn<String, TableRow>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext c)
+                      throws Exception {
+                    TableSchema schema = BigQueryHelpers.fromJsonString(
+                        c.sideInput(schemaView), TableSchema.class);
+                    String jobUuid = c.sideInput(jobIdTokenView);
+                    BigQuerySourceBase source = createSource(jobUuid);
+                    List<BoundedSource<TableRow>> sources =
+                    source.createSources(ImmutableList.of(
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: Indentation is somewhat off throughout the PR, I can fix myself when merging.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java [hunk 20]
```diff
@@ -79,6 +82,25 @@ public void processElement(ProcessContext c) {
...
-    abstract void cleanup(PipelineOptions options) throws Exception;
+    abstract void cleanup(ContextContainer container) throws Exception;
+  }
+
+  static class ContextContainer {
+    private PCollectionView<String> view;
+    private DoFn.ProcessContext context;
+
+    public ContextContainer(
+        DoFn.ProcessContext context, PCollectionView<String> view) {
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/PassThroughThenCleanup.java]: This and the type of line 90 should be `DoFn<?, ?>` - that'll get rid of the cast. I'll fix myself when merging.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 21]
```diff
@@ -400,7 +400,17 @@ public void testBuildSourceWithTableAndSqlDialect() {
   }
 
   @Test
-  public void testReadFromTable() throws IOException, InterruptedException {
+  public void testReadFromTableOldSource() throws IOException, InterruptedException {
+    testReadFromTable(false);
```
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Does this PR apply only to reading from a table, or from a query too?

**Comment by sammcveety** [comment]
Done, thank you!

**Comment by afcastano** [comment]
Hello!
Is this pull request merged into 2.1.0 release?

I can not find: setWithTemplateCompatibility method on my BigQueryIO Read

**Comment by kennknowles** [comment]
It is not, I'm afraid. It is merged to `master` and will be in the 2.2.0 release.

**Comment by afcastano** [comment]
:( Thanks,
Any timeframe for the 2.2.0 release?

**Comment by kennknowles** [comment]
We can't really predict releases with accuracy; it is governed by the community process. But our intention is to have a monthly cadence. Since 2.1.0 was significantly delayed, we already have a month's worth of changes to release in 2.2.0. In other words: very soon.

**Comment by afcastano** [comment]
:D great! Thanks!
